README.md
# tiddlysaver-python

[![GitHub last commit](https://img.shields.io/github/last-commit/itsamenathan/tiddlysaver-python)](https://github.com/itsamenathan/tiddlysaver-python) [![Docker Automated build](https://img.shields.io/docker/cloud/automated/itsamenathan/tiddlysaver-python)](https://hub.docker.com/r/itsamenathan/tiddlysaver-python) ![Docker Automated build](https://img.shields.io/docker/image-size/itsamenathan/tiddlysaver-python?sort=semver)

This is a python server for serving and saving tiddlywiki html files.  It will
host files from its current directory.  It will also save files there as well.

## Features

* Host tiddlywiki html files.
* Acts as a saver for tiddlywiki html files.
* Creates backups before saving files.
* Removes old backups using [python-rotate-backups](https://github.com/xolox/python-rotate-backups). Backup plan uses the `relaxed` flag.
  * Saves 10 minutely backups
  * Saves 24 hourly backups
  * Saves 30 daily backups
  * Saves 12 monthly backups
  * Saves 7 yearly backups
* Uses [flask](https://flask.palletsprojects.com/) web framework

## Running

### Docker

The image will run tiddlysaver from `/tiddlywiki`.

```bash
docker run -ti --rm \
           -e GUNICORN_CMD_ARGS="--bind=0.0.0.0:8000 --access-logfile=-" \
           -p 8000:8000 \
           -v $PWD/tiddlywiki:/tiddlywiki \
           itsamenathan/tiddlysaver-python:latest
```

### Python

```bash
pip install .
gunicorn -b 0.0.0.0:8000  "tiddlysaver:create_app()"
```

## Development

### Running

```bash
pip install -e .
export FLASK_APP=tiddlysaver
export FLASK_ENV=development
flask run --host 0.0.0.0 --port 8000
```

### Linting

```bash
docker run -e RUN_LOCAL=true -v ${PWD}:/tmp/lint github/super-linter
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.2] - 2020-09-04

### Changed

- Using ThreadingHTTPServer

## [1.0.1] - 2020-09-01

### Fixed

- Issue with corrupt overwrite.

### Changed

- Backup format to make it easier to read.
- Removed seconds from backup.  It wasn't useful with roteat_backup.

## [1.0.0] - 2020-08-29

### Added

- tiddlyserver.py
- Dockerfile


tiddlysaver/__init__.py
""" python server for serving and saving tiddlywiki html """
import datetime
import os
import shutil

from flask import Flask, request
from flask_autoindex import AutoIndex

import rotate_backups


def create_app():
    """ Create Flask app """
    app = Flask(__name__)

    def makebackup(src):
        app.logger.info("Making Backups")
        src = f".{src}"
        (srcpath, srcfile) = os.path.split(src)
        (srcname, src_ext) = os.path.splitext(srcfile)

        now = datetime.datetime.now().strftime("%Y-%m-%d-%H:%M")
        backup_dir = f"{srcpath}/backups/{srcname}"
        backup_file = f"{backup_dir}/{srcname}-{now}{src_ext}"

        if not os.path.exists(backup_dir):
            os.makedirs(backup_dir)
        shutil.copyfile(src, backup_file)

        return backup_file

    def cleanbackups(backup_dir):
        app.logger.info("Cleaning Backups")
        location = rotate_backups.coerce_location(backup_dir)
        backup = rotate_backups.RotateBackups(
            {"minutely": 10, "hourly": 24, "daily": 30, "monthly": 12, "yearly": 7},
            strict=False,
        )
        backup.rotate_backups(location)

    @app.route("/<path:file_path>", methods=["OPTIONS"])
    def _options(file_path):  # pylint: disable=unused-argument
        app.logger.info("Received OPTIONS")
        return "", 200, {"dav": "tw5/put"}

    @app.route("/<path:file_path>", methods=["PUT"])
    def _put_file(file_path):
        app.logger.info("Received PUT")

        data = request.get_data()

        # Don't save if we inside a backup directory
        if "backups/" in request.path:
            app.logger.warning("Do not save into a backup directory")
            return "Do not save into a backup directory", 403

        backup_file = makebackup(request.path)
        cleanbackups(os.path.dirname(backup_file))
        with open(file_path, "wb") as dst:
            dst.write(data)
        return "OK"

    AutoIndex(app, browse_root=os.path.curdir)

    return app


tiddlysaver/tiddlysaver.py
import datetime
import os
import shutil
import sys

from flask import Flask, request
from flask_autoindex import AutoIndex

import click

import rotate_backups

app = Flask(__name__)


def makebackup(src):
    src = f".{src}"
    (srcpath, srcfile) = os.path.split(src)
    (srcname, src_ext) = os.path.splitext(srcfile)

    now = datetime.datetime.now().strftime("%Y-%m-%d-%H:%M")
    backup_dir = f"{srcpath}/backups/{srcname}"
    backup_file = f"{backup_dir}/{srcname}-{now}{src_ext}"

    if not os.path.exists(backup_dir):
        os.makedirs(backup_dir)
    shutil.copyfile(src, backup_file)

    return backup_file


def cleanbackups(backup_dir):
    location = rotate_backups.coerce_location(backup_dir)
    backup = rotate_backups.RotateBackups(
        {"minutely": 10, "hourly": 24, "daily": 30, "monthly": 12, "yearly": 7},
        strict=False,
    )
    backup.rotate_backups(location)


@app.route("/<path:file_path>", methods=["OPTIONS"])
def options(file_path): # pylint: disable=unused-argument
    return "", 200, {"dav": "tw5/put"}


@app.route("/<path:file_path>", methods=["PUT"])
def put_file(file_path):
    data = request.get_data()

    # Don't save if we inside a backup directory
    if "backups/" in request.path:
        print("ERROR: Do not save into a backup directory")
        return "Do not save into a backup directory", 403

    backup_file = makebackup(request.path)
    cleanbackups(os.path.dirname(backup_file))
    with open(file_path, "wb") as dst:
        dst.write(data)
    return "OK"


@click.command()
@click.option(
    "-b",
    "--bind",
    default="127.0.0.1",
    envvar="BIND_HOST",
    metavar="ADDRESS",
    help="Specify alternate bind address [default: 127.0.0.0]",
)
@click.option(
    "-p",
    "--port",
    default=8000,
    envvar="BIND_PORT",
    help="Specify alternate port [default: 8000]",
)
def http_serve(bind, port):
  """Script to serve a tiddlywiki saver."""
  # AutoIndex takes care of GET routes on / and /<path:file_path>
  AutoIndex(app, browse_root=os.path.curdir)
  app.run(host=bind, port=port)
